src/wolframite/runtime/system.clj

```diff
@@ -0,0 +1,161 @@
+(ns wolframite.runtime.system
+  "For dealing with the runtime system, primarily the operating system specific defaults."
+  (:require
+   [babashka.fs :as fs]
+   [clojure.string :as string])
+  (:import
+   (java.lang System)))
+
+(def supported-OS #{:linux :mac :windows})
+
+;; Default path information can be listed here, in order of preference.
+(def ^:private defaults
+  [{:root  "/usr/local/Wolfram/Mathematica"
+    :kernel "Executables/MathKernel"
+    :product :mathematica
+    :os :linux}
+   {:root  "/opt/Mathematica"
+    :kernel "Executables/MathKernel"
+    :product :mathematica
+    :os :linux}
+   {:root  "/usr/local/Wolfram/WolframEngine"
+    :kernel "Executables/WolframKernel"
+    :product :wolfram-engine
+    :os :linux}
+
+   {:root  "/Applications/Mathematica.app/Contents"
+    :kernel "MacOS/MathKernel"
+    :product :mathematica
+    :os :mac}
+   {:root   "/Applications/Wolfram Engine.app/Contents/Resources/Wolfram Player.app/Contents"
+    :kernel "MacOS/WolframKernel"
+    :product :wolfram-engine
+    :os :mac}
+
+   {:root "/c:/Program Files/Wolfram Research/Mathematica/"
+    :kernel "WolframKernel.exe"
+    :product :mathematica
+    :os :windows}
+   {:root "/c:/Program Files/Wolfram Research/Wolfram Engine/"
+    :kernel "MathKernel.exe"
+    :product :wolfram-engine
+    :os :windows}])
+
+(defn- user-paths
+  "The paths given via Java property or environment variables.
+
+  NOTE: These should be root directories that distinguish Mathematica or Wolfram engine files, not subdirectories. E.g.  '/usr/local/Wolfram/Mathematica'.
+
+  TODO:
+  - (jh) alternatively, support Java system properties?"
+  []
+  (let [names  ["JLINK_JAR_PATH"
+                "MATHEMATICA_INSTALL_PATH"
+                "WOLFRAM_INSTALL_PATH"]
+        method (fn [f] (-> f (nth  2) first str (string/split #"/") last))
+        fs ['(fn [name] (System/getProperty name)) '(fn [name] (System/getenv name))]]
+    (->> (map (fn [f]
+                (map (fn [name] [[(keyword name) ((eval f) name)]
```

Comment:
pls don't use eval, that is just evil. No need for that. I struggle a little to see what this code tries to do, so I cannot recommend a better approach now. But I am sure we can do st. simpler.

Reply:
I'm sure you're right here. I may have tried too hard (or not hard enough!) to condense this code. I'll have a go at making it less evil (more Holy ;) ). The basic idea is to take a list of variables and to check if they were supplied at the property or environment variable level and to collect the results in a map.